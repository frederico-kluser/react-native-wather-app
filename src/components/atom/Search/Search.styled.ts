import styled from 'styled-components/native';
import {
  BORDER_RADIUS,
  INPUT_HEIGHT,
  ITEM_MARGIN,
  GLOBAL_PADDING,
  SHADOW_EFFECT,
} from '../../../styles/theme';

export const Container = styled.View`
  ${SHADOW_EFFECT}
  border-radius: ${BORDER_RADIUS}px;
  margin-bottom: ${ITEM_MARGIN}px;
  width: 100%;
  padding-left: ${GLOBAL_PADDING}px;
  padding-right: ${INPUT_HEIGHT}px;
`;

export const InputSearch = styled.TextInput`
  height: ${INPUT_HEIGHT}px;
  width: 100%;
`;

export const Option = styled.TouchableOpacity`
  left: ${GLOBAL_PADDING}px;
  position: relative;
  overflow: hidden;
`;

export const OptionText = styled.Text`
  height: ${INPUT_HEIGHT}px;
  line-height: ${INPUT_HEIGHT}px;
`;

export const SearchButton = styled.TouchableOpacity`
  position: absolute;
  right: -${GLOBAL_PADDING}px;
  top: -${GLOBAL_PADDING}px;
`;

const iconSpace = 8;
export const SearchIcon = styled.Image`
  position: absolute;
  right: ${GLOBAL_PADDING + iconSpace / 2}px;
  top: ${GLOBAL_PADDING + iconSpace / 2}px;
  width: ${INPUT_HEIGHT - iconSpace}px;
  height: ${INPUT_HEIGHT - iconSpace}px;
`;

export default Container;
