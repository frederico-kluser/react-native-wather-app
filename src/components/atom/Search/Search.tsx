import React, {useEffect} from 'react';
import getCityNames from '../../../api/cityNames';
import getCityWeatherInfo, {
  getCityWeatherNextDaysInfo,
} from '../../../api/cityWeatherInfo';
import convertKelvinToCelsius from '../../../helpers/conversor';
import {closeIcon} from '../../../styles/icons';
import Container, {
  InputSearch,
  Option,
  OptionText,
  SearchButton,
  SearchIcon,
} from './Search.styled';

const Search = ({
  cities,
  inputFilter,
  options,
  setCities,
  setInputFilter,
  setOptions,
}: any) => {
  const handleInput = async (index = 0) => {
    if (options.length) {
      const data: any = await getCityWeatherInfo(
        options[index].structured_formatting.main_text,
      );
      const data2 = await getCityWeatherNextDaysInfo(
        data.coord.lat,
        data.coord.lon,
      );

      const item = {
        city: data.name,
        country: data.sys.country,
        description: options[index].description,
        favorite: false,
        humidity: parseInt(data.main.humidity),
        maxTemperature: convertKelvinToCelsius(data.main.temp_max),
        minTemperature: convertKelvinToCelsius(data.main.temp_min),
        nextDays: data2.daily,
        status: data.weather[0].description,
        temperature: convertKelvinToCelsius(data.main.temp),
        wind: parseInt(data.wind.speed),
      };
      console.log(data, 'data');
      console.log(item, 'item');
      setCities((prevState: any[]) => {
        return [...prevState, item];
      });
    }
    setInputFilter('');
  };

  useEffect(() => {
    if (inputFilter) {
      getCityNames(inputFilter).then(({predictions}: any) => {
        setOptions(predictions.map((item: any) => item));
      });
    } else {
      setOptions([]);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [inputFilter]);

  return (
    <Container>
      <InputSearch
        placeholder="Digite o nome da cidade"
        value={inputFilter}
        onChangeText={setInputFilter}
        onSubmitEditing={() => handleInput()}
        quandity={options.length}
      />
      {options.map(({description}: any, index: number) => (
        <Option key={description} onPress={() => handleInput(index)}>
          <OptionText>{description}</OptionText>
        </Option>
      ))}
      {!!inputFilter && (
        <SearchButton onPress={() => setInputFilter('')}>
          <SearchIcon source={closeIcon} />
        </SearchButton>
      )}
    </Container>
  );
};

export default Search;
