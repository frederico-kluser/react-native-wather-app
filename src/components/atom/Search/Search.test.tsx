import React from 'react';
import renderer from 'react-test-renderer';
import Search from './Search';

jest.useFakeTimers();

test('renders correctly', () => {
  const tree = renderer.create(<Search options={[]} />).toJSON();
  expect(tree).toMatchSnapshot();
});
