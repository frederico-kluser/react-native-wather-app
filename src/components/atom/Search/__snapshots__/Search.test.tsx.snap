// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 16,
        "marginBottom": 32,
        "paddingLeft": 16,
        "paddingRight": 48,
        "shadowColor": "rgba(0, 0, 0, 0.24)",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 8,
        "width": "100%",
      },
    ]
  }
>
  <TextInput
    placeholder="Digite o nome da cidade"
    style={
      Array [
        Object {
          "height": 48,
          "width": "100%",
        },
      ]
    }
  />
</View>
`;
